File: KIOP_WS1415/Blatt02/d_Observer_JavaScript/Observer.js
var Observable = function() {
  obs = [];

  this.attach = function(Observer) {
    obs.push(Observer);
  };

  this.detach = function(Observer) {
    for(var i = obs.length; i--;) {
      if(obs[i] === Observer) {
        obs.splice(i, 1);
      }
    }
  };

  this.notify = function() {
    for(var i = obs.length; i--;) {
      obs[i].update();
    }
  };

  return this;
};

var Subject = function(value) {
  this.value = value;

  this.setVal = function (val) {
    this.value = val;
    this.notify();
  };

  return this;
};

Observable.call(Subject.prototype);

var Observer = function(subject) {
  this.sub = subject;
  this.sub.attach(this);
  this.update = function() { };
  return this;
};

var HexObserver = function(subject) {
  var instance = new Observer(subject);

instance.update = function() {
    var v = this.sub.value;
    console.log(v + " als Hex ist "+ v.toString(16));
  };

  return instance;
};

var OctObserver = function(subject) {
  var instance = new Observer(subject);

instance.update = function() {
    var v = this.sub.value;
    console.log(v + " als Oct ist "+ v.toString(8));
  };

  return instance;
};

var sub = new Subject(10);
var obs1 = new HexObserver(sub);
var obs2 = new OctObserver(sub);
var obs3 = new HexObserver(sub);
var obs4 = new OctObserver(sub);

sub.setVal(15);
sub.detach(obs1);
sub.detach(obs3);
sub.setVal(10);
sub.setVal(5);

File: KIOP_WS1415/Blatt02/a_Observer/Observer.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include "Observer.h"

void Subject::setVal(int val) {
  value = val;
	notify();
}

template <class Super> 
void ObsMixIn<Super>::attach(cheatObs *o) {
	obs.push_back(o);
}

template <class Super>
void ObsMixIn<Super>::detach(cheatObs *o) {
    obs.erase(remove(obs.begin(), obs.end(), o), obs.end());
}

template <class Super>
void ObsMixIn<Super>::notify() {
  for (int i = 0; i < (signed) obs.size(); i++)
    obs[i]->update();
}

template <class T>
Observer<T>::Observer(ObsMixIn<T> *mod) {
    model = mod;
    model->attach(this);
}

void HexObserver::update() {
    int v = getSubject()->getVal();
    cout << v << " als Hex ist " << std::hex << v << std::dec << '\n';
}

void OctObserver::update() {
    int v = getSubject()->getVal();
    cout << v << " als Octal ist " << std::oct << v << std::dec << '\n';
}

int main() {
  ObsMixIn<Subject> subj;
  HexObserver hexObs(&subj);
  OctObserver octObs(&subj);
  subj.setVal(14);
  subj.detach(&hexObs);
  subj.setVal(64);
  subj.setVal(100);
}
